ops/pebble.py

```diff
@@ -752,5 +872,214 @@ def get_services(self, names: List[str] = None) -> List[ServiceInfo]:
+
+    @staticmethod
+    def _raise_on_path_error(resp, path):
+        result = resp['result'] or []  # in case it's null instead of []
+        paths = {item['path']: item for item in result}
+        if path not in paths:
+            raise ProtocolError('path not found in response metadata: {}'.format(resp))
+        error = paths[path].get('error')
+        if error:
+            raise PathError(error['kind'], error['message'])
+
+    def write_content(
+            self, path: str, content: typing.Union[bytes, str], encoding: str = 'utf-8',
+            make_dirs: bool = False, permissions: int = None, user: typing.Union[str, int] = None,
+            group: typing.Union[str, int] = None):
+        """Write content to a given file path on the remote system.
+
+        Args:
+            path: Path of the file to write to on the remote system.
+            content: Content to write (str or bytes).
+            encoding: Encoding to use for encoding content str to bytes.
+                Ignored if content is bytes.
+            make_dirs: If True, create parent directories if they don't exist.
+            permissions: Permissions (mode) to create file with (Pebble default
+                is 0o644).
+            user: UID or username for file.
+            group: GID or group name for file.
+        """
+        info = self._make_auth_dict(permissions, user, group)
+        info['path'] = path
+        if make_dirs:
+            info['make-dirs'] = True
+        metadata = {
+            'action': 'write',
+            'files': [info],
+        }
+
+        multipart = MIMEMultipart('form-data')
+
+        part = MIMEBase('application', 'json')
+        part.add_header('Content-Disposition', 'form-data', name='request')
+        part.set_payload(json.dumps(metadata))
+        multipart.attach(part)
+
+        part = MIMEBase('application', 'octet-stream')
+        part.add_header('Content-Disposition', 'form-data', name='files', filename=path)
+        if isinstance(content, str):
+            content = content.encode(encoding)
+        part.set_payload(content)
+        multipart.attach(part)
+
+        data = multipart.as_bytes()  # must be called before accessing multipart['Content-Type']
+        headers = {
+            'Accept': 'application/json',
+            'Content-Type': multipart['Content-Type'],
+        }
+        response = self._request_raw('POST', '/v1/files', None, headers, data)
+        self._ensure_content_type(response.headers, 'application/json')
+        resp = _json_loads(response.read())
+        self._raise_on_path_error(resp, path)
+
+    @staticmethod
+    def _make_auth_dict(permissions, user, group) -> typing.Dict:
+        d = {}
+        if permissions is not None:
+            d['permissions'] = format(permissions, '03o')
+        if user is not None:
+            if isinstance(user, int):
+                d['user-id'] = user
+            elif isinstance(user, str):
+                d['user'] = user
```

Comment:
What if we want to send/receive both a uid and a username? Seems safer to be explicit.

Same for groups below.

Reply:
The issue for sending both uid and username is that I believe that creates ambiguity. Did you mean 1001 or user 'foo' who is 1002?

Reply:
The answers to these specific questions seem easy: I meant user 'foo' with uid 1001, and 1002 is whoever 1002 is, of course. Those things are an inherent part of the underlying problem you are dealing with: there are _really_ two systems with their own data. So the more interesting question is what you do with that information.

For example, if we are transferring user 'foo' that has id 1001 to the local system, and neither foo nor 1001 exist locally and you want to create them, it's nice to replicate the exact data. If user 'foo' already exists, the better behavior seems to be preserving the username and warning that the uid diverges. We might also opt to forbid it entirely, but we can only do that if we have the data, and we might even get started by having both parameters and enforcing the use of a single one of them for now.

But if you have both uid and username under the same parameter, game over. The API was cooked in a way that you cannot even decide to forbid or catch mistakes, or evolve in a polished way, because we've encouraged people to use a lesser API upfront.

Reply:
If updated the `write_content` and `make_dir` functions signatures to accept both `user_id` and `user`, `group_id` and `group`. The current behavior (and this matches the underlying HTTP API) is if both are set, the IDs take precedence. Having both args also makes the types a bit clearer and the signature a bit more direct. The idea is people will use one or the other.